File: apiwrapper/src/main/java/com/pinkertone/apiwrapper/ApiWrapper.java
package com.pinkertone.apiwrapper;

import com.pinkertone.apiwrapper.types.Token;
import com.pinkertone.apiwrapper.types.User;
import com.pinkertone.apiwrapper.types.UserProfile;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class ApiWrapper {
    private ApiService apiServiceSingleton;
    private String authToken = "";
    private String baseUrl;
    private String language;

    public void setAuthToken(String authToken) {
        this.authToken = authToken;
        apiServiceSingleton.setAuthToken(authToken);
    }

    public ApiWrapper(String baseUrl, String authToken, String language){
        this.baseUrl = baseUrl;
        this.authToken = authToken;
        this.language = language;
        this.apiServiceSingleton = ApiService.getInstance(baseUrl, authToken, language);
    }

    public ApiWrapper(String baseUrl, String language) {
        this.baseUrl = baseUrl;
        this.language = language;
        this.apiServiceSingleton = ApiService.getInstance(baseUrl, language);
    }

    public void loginUser(final ICallback<Token> callback, String username, String password) {
        Call<Token> loginCall = apiServiceSingleton.apiService
                .loginUser(username, password);

        loginCall.enqueue(new Callback<Token>() {
            @Override
            public void onResponse(Call<Token> call, Response<Token> response) {
                callback.onSuccess(response.body());
                authToken = response.body().token;
                apiServiceSingleton.setAuthToken(authToken);
            }

            @Override
            public void onFailure(Call<Token> call, Throwable t) {
                t.printStackTrace();
                callback.onFail(t);
            }
        });
    }

    public void createUser(final ICallback<Token> callback, String username, String email, String password) {
        Call<Token> createCall = apiServiceSingleton.apiService
                .createUser(username, email, password);

        createCall.enqueue(new Callback<Token>() {
            @Override
            public void onResponse(Call<Token> call, Response<Token> response) {
                callback.onSuccess(response.body());
                authToken = response.body().token;
                apiServiceSingleton.setAuthToken(authToken);
            }

            @Override
            public void onFailure(Call<Token> call, Throwable t) {
                t.printStackTrace();
                callback.onFail(t);
            }
        });
    }

    public void getUserInfo(final ICallback<UserProfile> callback) {
        final Call<UserProfile> userInfoCall = apiServiceSingleton.apiService
                .getUserInfo();
        userInfoCall.enqueue(new Callback<UserProfile>() {
            @Override
            public void onResponse(Call<UserProfile> call, Response<UserProfile> response) {
                callback.onSuccess(response.body());
            }

            @Override
            public void onFailure(Call<UserProfile> call, Throwable t) {
                t.printStackTrace();
            }
        });

    }

    public boolean isAuthorized() {
        return !authToken.isEmpty() & !apiServiceSingleton.getAuthToken().isEmpty();
    }
}

File: apiwrapper/src/main/java/com/pinkertone/apiwrapper/ApiWrapper.kt
package com.pinkertone.apiwrapper

import com.pinkertone.apiwrapper.types.Token
import com.pinkertone.apiwrapper.types.UserProfile

// Singleton for client to make requests
class ApiWrapper private constructor(
    private val baseUrl: String,
    private val language: String,
) {
    private val apiService: ApiService = ApiService(baseUrl, language)
    private val responseHandler: ResponseHandler = ResponseHandler()

    fun setAuthToken(authToken: String?) {
        apiService.setAuthToken(authToken ?: "")
    }

    private constructor(
        baseUrl: String,
        authToken: String = "",
        language: String
    ) : this(baseUrl, language) {
        apiService.setAuthToken(authToken)
    }

    suspend fun loginUser(username: String, password: String): Resource<Token> {
        return try {
            val token = apiService.api.loginUser(username, password)
            apiService.setAuthToken(token.string)
            responseHandler.handleSuccess(token)
        } catch (e: Exception) {
            responseHandler.handleException(e)
        }
    }

    suspend fun createUser(
        username: String,
        email: String,
        password: String
    ): Resource<Token> {
        return try {
            responseHandler.handleSuccess(
                apiService.api.createUser(username, email, password)
            )
        } catch (e: Exception) {
            responseHandler.handleException(e)
        }
    }

    suspend fun getUserInfo(): Resource<UserProfile> {
        return try {
            responseHandler.handleSuccess(apiService.api.getUserInfo())
        } catch (e: Exception) {
            responseHandler.handleException(e)
        }
    }

    val isAuthorized: Boolean
        get() = apiService.getAuthToken().isNotEmpty()

    companion object {
        private var instance: ApiWrapper? = null

        @Synchronized
        fun getInstance(baseUrl: String, authToken: String, language: String): ApiWrapper {
            if (instance == null) {
                instance = ApiWrapper(baseUrl, authToken, language)
            }
            return instance!!
        }

        @Synchronized
        fun getInstance(baseUrl: String, language: String): ApiWrapper {
            if (instance == null) {
                instance = ApiWrapper(baseUrl, language)
            }
            return instance!!
        }

        @Synchronized
        @Throws(RuntimeException::class)
        fun getInstance(): ApiWrapper {
            if (instance == null) {
                throw RuntimeException("ApiService was not initialized before")
            }
            return instance!!
        }
    }

}